Makes interaction region visible on a HTMLTextFormControlElement when focused.

amarchenko opened the pull request:
#### c31a3bd6fa34624051e51534ba90c847f98b6488
<pre>
Makes interaction region visible on a HTMLTextFormControlElement when focused.
<a href="https://bugs.webkit.org/show_bug.cgi?id=275064">https://bugs.webkit.org/show_bug.cgi?id=275064</a>
<a href="https://rdar.apple.com/120749461">rdar://120749461</a>

Reviewed by Aditya Keerthi.

Interaction region is visible in almost all states of a HTMLTextFormControlElement
except when editing the text and when focused + edited + has-value.

* LayoutTests/interaction-region/text-input-focused-edited-empty-expected.txt: Added.
* LayoutTests/interaction-region/text-input-focused-edited-empty.html: Added.
* LayoutTests/interaction-region/text-input-focused-edited-expected.txt: Copied from LayoutTests/interaction-region/text-input-focused-expected.txt.
* LayoutTests/interaction-region/text-input-focused-edited.html: Added.
* LayoutTests/interaction-region/text-input-focused-expected.txt:
* LayoutTests/interaction-region/textarea-focused-edited-empty-expected.txt: Added.
* LayoutTests/interaction-region/textarea-focused-edited-empty.html: Added.
* LayoutTests/interaction-region/textarea-focused-edited-expected.txt: Copied from LayoutTests/interaction-region/text-input-focused-expected.txt.
* LayoutTests/interaction-region/textarea-focused-edited.html: Added.
* LayoutTests/interaction-region/textarea-focused-expected.txt:
* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::shouldAllowNonInteractiveCursorForElement):

Canonical link: <a href="https://commits.webkit.org/279828@main">https://commits.webkit.org/279828@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2459068e8a9aae8d58a9e4f7e95a2b53bb95ddd2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54642 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34064 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7217 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57922 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5375 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56943 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41603 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5383 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44263 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3627 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56737 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32205 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47329 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25388 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29002 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4663 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3515 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/50747 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4877 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59512 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29880 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5023 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51687 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31025 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47415 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51071 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11984 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32012 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30812 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

etiennesegonzac commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -162,8 +162,8 @@ static bool shouldAllowNonInteractiveCursorForElement(const Element& element)
-    if (is<HTMLTextFormControlElement>(element))
-        return !element.focused();
+    if (auto* textElement = dynamicDowncast<HTMLTextFormControlElement>(element))
+        return !textElement->focused() || !textElement->lastChangeWasUserEdit() || textElement->value().isEmpty();
```
You should be able to test the new logic here with `UIHelper.activateAndWaitForInputSession*()` and `UIHelper.typeCharacter()`. We still want to cover cases where we hide the region!

amarchenko replied:
Added 2 new tests:  `textarea-focused-edited.html` and `text-input-focused-edited.html`.

pxlcoder commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -162,8 +162,8 @@ static bool shouldAllowNonInteractiveCursorForElement(const Element& element)
 
-    if (is<HTMLTextFormControlElement>(element))
-        return !element.focused();
+    if (auto* textElement = dynamicDowncast<HTMLTextFormControlElement>(element))
```
```suggestion
    if (RefPtr textElement = dynamicDowncast<HTMLTextFormControlElement>(element))
```

amarchenko replied:
@pxlcoder all other `dynamicDowncast`s in the file use variations of `auto` as type for the returned result. Is there any particular reason why mine should be different?

pxlcoder replied:
We are adopting smart pointers everywhere in WebKit, so we shouldn't introduce new raw pointer usage. The other cases were likely before the policy was announced, or were simply missed during review.

https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines

etiennesegonzac reviewed (commented):
It is great to see these tests!

I think we can cover the last part of the logic (`textElement->value().isEmpty()`), by adding similar tests where we call `document.execCommand("Delete")` after typing.
(Wouldn't block on it but worth trying.)

And let's make sure they're included in the commit message :)

etiennesegonzac reviewed (commented):
Great! Ready for the official review by @pxlcoder :)

amarchenko committed:
Makes interaction region visible on a HTMLTextFormControlElement when focused.
https://bugs.webkit.org/show_bug.cgi?id=275064
rdar://120749461

Reviewed by Aditya Keerthi.

Interaction region is visible in almost all states of a HTMLTextFormControlElement
except when editing the text and when focused + edited + has-value.

* LayoutTests/interaction-region/text-input-focused-edited-empty-expected.txt: Added.
* LayoutTests/interaction-region/text-input-focused-edited-empty.html: Added.
* LayoutTests/interaction-region/text-input-focused-edited-expected.txt: Copied from LayoutTests/interaction-region/text-input-focused-expected.txt.
* LayoutTests/interaction-region/text-input-focused-edited.html: Added.
* LayoutTests/interaction-region/text-input-focused-expected.txt:
* LayoutTests/interaction-region/textarea-focused-edited-empty-expected.txt: Added.
* LayoutTests/interaction-region/textarea-focused-edited-empty.html: Added.
* LayoutTests/interaction-region/textarea-focused-edited-expected.txt: Copied from LayoutTests/interaction-region/text-input-focused-expected.txt.
* LayoutTests/interaction-region/textarea-focused-edited.html: Added.
* LayoutTests/interaction-region/textarea-focused-expected.txt:
* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::shouldAllowNonInteractiveCursorForElement):

Canonical link: https://commits.webkit.org/279828@main
